Add text component

EternityForest opened the pull request:
# What does this implement/fix?
This is an updated version of the text inputs component branch, building on work by [mauritskorse](https://github.com/mauritskorse) in https://github.com/esphome/esphome/pull/3638 which seems to have not gotten any updates for a few months.

Almost all of the work here is his, and he deserves full credit, I have only solved some merge conflicts and made a few minor changes as mentioned in the review, as I had some free time and he seems to be busy lately.

The original description is here:

> Adding a input text component (similar to [Home Assistant](https://www.home-assistant.io/integrations/input_text/)). It allows for setting a string by the user through webserver comp front-end (v1). Implementation for v2 webserver front-end and Home Assistant is not done yet.


But I have since added V2 support(https://github.com/esphome/esphome-webserver/pull/43), and I believe that I have covered most of the requested changes in the original.

Changes relative to the original:

Synced to upstream

- Changed the message type numbers as the old ones conflicted with newer upstream work
- Added the ability to persist strings to flash
- Use this-> for property references instead of the implicit ones
- add validate_ method(I wasn't quite sure what was actually needed or how validate works, more of a draft there, maybe @jesserockz can weigh in on if I did it right?)
- Deleted commented out code that did not appear to be relevant
- value_ begins at nullpointer by default(Not sure if that's actually a thing, or if the other components do it for a reason?)
- Added min_length, max_length, and pattern to the JSON in the web server.


The protocol buffers here were **not** compiled with 3.6 as mentioned in a comment, because that version is not available in newer Ubuntus.  If it's an exact version requirement and not a "Any later minor version" kind of thing, I can set up a VM and fix this.

Replaces and closes #3638

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3293

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
  - platform: template
    name: "test 2 text"
    id: "it_2"
    icon: "mdi:text"
    optimistic: true
    internal: false
    initial_value: "Welcome ESPHOME"
    restore_value: true
    max_restore_data_length: 24

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

(The original PR is marked as having documentation, but I have not updated it to include the newly-fixed restore_value)

mauritskorse committed:
initial add text_input component

mauritskorse committed:
fixed additional files to implement text_input component

mauritskorse committed:
Adding text input component to MQTT en Template comps

mauritskorse committed:
minor fixes

mauritskorse committed:
more fixes

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
Renaming from text_input to text_input in line with HA component

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
added password support to web v1.0 front-end

mauritskorse committed:
Add support for copying input_text comp

mauritskorse committed:
adding codeowner

mauritskorse committed:
adding component tests

mauritskorse committed:
add to codeowner defs

mauritskorse committed:
fixing tests

mauritskorse committed:
fixing tests results

mauritskorse committed:
fixing tests results

mauritskorse committed:
fixing tests results

mauritskorse committed:
fixing tests results

mauritskorse committed:
Merge remote-tracking branch 'origin/add-text-input-component' into add-text-input-component

mauritskorse committed:
further fixes

mauritskorse committed:
further fixes

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
copy support for input_text comp

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
fixing lint issues after rebase

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
lint fix trailing spaces

mauritskorse committed:
Merge branch 'esphome:dev' into add-text-input-component

mauritskorse committed:
Merge branch 'esphome:dev' into add-text-input-component

mauritskorse committed:
renaming from input_text to text in line with HA 2022.12

mauritskorse committed:
.

mauritskorse committed:
Merge branch 'dev' into add-text-input-component

mauritskorse committed:
rename text mode from STRING to TEXT

mauritskorse committed:
Merge branch 'esphome:dev' into add-text-input-component

mauritskorse committed:
pylint fix codeowners

mauritskorse committed:
pylint fix test

mauritskorse committed:
Merge branch 'add-text-input-component' of https://github.com/mauritskorse/esphome into add-text-input-component

mauritskorse committed:
fix and add support for automation on text component

mauritskorse committed:
.

mauritskorse committed:
fix setting proper ids

mauritskorse committed:
fix pylint suggestion

mauritskorse committed:
.

mauritskorse committed:
fix min max in line with HA text entity

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
.

mauritskorse committed:
fix webserver text input

mauritskorse committed:
.

mauritskorse committed:
Merge branch 'esphome:dev' into add-text-input-component

mauritskorse committed:
renaming min/max

mauritskorse committed:
Merge branch 'add-text-input-component' of https://github.com/mauritskorse/esphome into add-text-input-component

Commit:
Initial work on making text inputs savable

Commit:
Merge remote-tracking branch 'upstream/dev' into sync-add-text-input-component

Move the message codes up to 95,96, 97, because they conflicted with new
bluetooth stuff.

mreditor97 in CODEOWNERS kept as lowercase.

Commit:
Rebuild protobufs(Note: 3.12 not 3.6, not sure if thats bad)

Commit:
Format and fix array size issue

Commit:
Add the validation

Commit:
Merge branch 'add-text-input-component-prefs' into sync-add-text-input-component

Commit:
Merge branch 'text-input-cleanup' into sync-add-text-input-component

Commit:
Add text traits to the webserver JSON

Commit:
Fix linting errors, move CONF_SET_ACTION to const.py, fix tests

Commit:
Remove YAML whitespace

EternityForest committed:
Fix linter styling errors in template_text.h

Commit:
63 char limit on restoring text, to match globals(among other things)

jesserockz commented on esphome/components/api/api.proto:
```diff
@@ -1517,3 +1518,49 @@ message AlarmControlPanelCommandRequest {
+
+// ===================== TEXT =====================
+enum TextMode {
+  TEXT_MODE_AUTO = 0;
```
I don't see how an auto mode will help at all..
HA doesn't seem to have this as a mode either

jesserockz commented on esphome/components/text/__init__.py:
```diff
@@ -0,0 +1,141 @@
+    "PASSWORD": TextMode.TEXT_MODE_PASSWORD,  # to be implemented for keys, passwords, etc.
+}
+
+icon = cv.icon
```
```suggestion
```

jesserockz commented on esphome/components/text/__init__.py:
```diff
@@ -0,0 +1,141 @@
+TextMode = text_ns.enum("TextMode")
+
+TEXT_MODES = {
+    "AUTO": TextMode.TEXT_MODE_AUTO,
```
```suggestion
```

jesserockz commented on esphome/components/text/__init__.py:
```diff
@@ -0,0 +1,141 @@
+                cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(TextStateTrigger),
+            }
+        ),
+        cv.Optional(CONF_MODE, default="AUTO"): cv.enum(TEXT_MODES, upper=True),
```
```suggestion
        cv.Required(CONF_MODE): cv.enum(TEXT_MODES, upper=True),
```

jesserockz commented on esphome/components/text/text_call.cpp:
```diff
@@ -0,0 +1,52 @@
+  }
+}
+
+void TextCall::perform() {
```
Missing the call to validate_ ?

jesserockz commented on esphome/components/text/text_traits.cpp:
```diff
@@ -0,0 +1,10 @@
+#include "esphome/core/log.h"
```
Not much point in this file...

jesserockz commented on esphome/components/text/text_traits.h:
```diff
@@ -0,0 +1,40 @@
+namespace text {
+
+enum TextMode : uint8_t {
+  TEXT_MODE_AUTO = 0,
+  TEXT_MODE_TEXT = 1,
+  TEXT_MODE_PASSWORD = 2,
```
```suggestion
  TEXT_MODE_TEXT = 0,
  TEXT_MODE_PASSWORD = 1,
```

jesserockz commented on tests/component_tests/text/test_text.yaml:
```diff
@@ -0,0 +1,31 @@
+esphome:
```
These tests are not used by CI. 
Please add to one of the test yaml files in the tests folder.

jesserockz reviewed (changes requested):
Overall looks fine. I have not tested yet, but will hopefully this week.
A few things below.

I don't think we need to have an auto mode (there is no way to determine if you want text or password), and mode should be required not optional.

EternityForest commented:
Yeah, I'm not exactly sure what the intent of the auto mode was, but I can
certainly get back on this PR and make those changes!

I could see the value of making generic text as the default and having mode
be optional, but then again, explicit is better than implicit.

I'm impressed by the thoroughness of the review and attention to code
quality!

On Mon, Oct 23, 2023, 9:15 PM Jesse Hills ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Overall looks fine. I have not tested yet, but will hopefully this week.
> A few things below.
>
> I don't think we need to have an auto mode (there is no way to determine
> if you want text or password), and mode should be required not optional.
> ------------------------------
>
> In esphome/components/api/api.proto
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369461731>:
>
> > @@ -1517,3 +1518,49 @@ message AlarmControlPanelCommandRequest {
>    AlarmControlPanelStateCommand command = 2;
>    string code = 3;
>  }
> +
> +// ===================== TEXT =====================
> +enum TextMode {
> +  TEXT_MODE_AUTO = 0;
>
> I don't see how an auto mode will help at all..
> HA doesn't seem to have this as a mode either
> ------------------------------
>
> In esphome/components/text/__init__.py
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369489897>:
>
> > +    "TextStateTrigger", automation.Trigger.template(cg.std_string)
> +)
> +
> +# Actions
> +TextSetAction = text_ns.class_("TextSetAction", automation.Action)
> +
> +# Conditions
> +TextMode = text_ns.enum("TextMode")
> +
> +TEXT_MODES = {
> +    "AUTO": TextMode.TEXT_MODE_AUTO,
> +    "TEXT": TextMode.TEXT_MODE_TEXT,
> +    "PASSWORD": TextMode.TEXT_MODE_PASSWORD,  # to be implemented for keys, passwords, etc.
> +}
> +
> +icon = cv.icon
>
> ⬇️ Suggested change
>
> -icon = cv.icon
>
> ------------------------------
>
> In esphome/components/text/__init__.py
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369490013>:
>
> > +Text = text_ns.class_("Text", cg.EntityBase)
> +TextPtr = Text.operator("ptr")
> +
> +# Triggers
> +TextStateTrigger = text_ns.class_(
> +    "TextStateTrigger", automation.Trigger.template(cg.std_string)
> +)
> +
> +# Actions
> +TextSetAction = text_ns.class_("TextSetAction", automation.Action)
> +
> +# Conditions
> +TextMode = text_ns.enum("TextMode")
> +
> +TEXT_MODES = {
> +    "AUTO": TextMode.TEXT_MODE_AUTO,
>
> ⬇️ Suggested change
>
> -    "AUTO": TextMode.TEXT_MODE_AUTO,
>
> ------------------------------
>
> In esphome/components/text/__init__.py
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369491063>:
>
> > +    "TEXT": TextMode.TEXT_MODE_TEXT,
> +    "PASSWORD": TextMode.TEXT_MODE_PASSWORD,  # to be implemented for keys, passwords, etc.
> +}
> +
> +icon = cv.icon
> +
> +TEXT_SCHEMA = cv.ENTITY_BASE_SCHEMA.extend(cv.MQTT_COMPONENT_SCHEMA).extend(
> +    {
> +        cv.OnlyWith(CONF_MQTT_ID, "mqtt"): cv.declare_id(mqtt.MQTTTextComponent),
> +        cv.GenerateID(): cv.declare_id(Text),
> +        cv.Optional(CONF_ON_VALUE): automation.validate_automation(
> +            {
> +                cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(TextStateTrigger),
> +            }
> +        ),
> +        cv.Optional(CONF_MODE, default="AUTO"): cv.enum(TEXT_MODES, upper=True),
>
> ⬇️ Suggested change
>
> -        cv.Optional(CONF_MODE, default="AUTO"): cv.enum(TEXT_MODES, upper=True),
> +        cv.Required(CONF_MODE): cv.enum(TEXT_MODES, upper=True),
>
> ------------------------------
>
> In esphome/components/text/text_call.cpp
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369494564>:
>
> > +  int sz = this->value_.value().size();
> +
> +  if (sz > this->parent_->traits.get_max_length()) {
> +    ESP_LOGW(TAG, "'%s' - Value set for TextCall is too long", name);
> +    this->value_.reset();
> +    return;
> +  }
> +
> +  if (sz < this->parent_->traits.get_min_length()) {
> +    ESP_LOGW(TAG, "'%s' - Value set for TextCall is too short", name);
> +    this->value_.reset();
> +    return;
> +  }
> +}
> +
> +void TextCall::perform() {
>
> Missing the call to validate_ ?
> ------------------------------
>
> In esphome/components/text/text_traits.cpp
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369495686>:
>
> > @@ -0,0 +1,10 @@
> +#include "esphome/core/log.h"
>
> Not much point in this file...
> ------------------------------
>
> In esphome/components/text/text_traits.h
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369496212>:
>
> > +  TEXT_MODE_AUTO = 0,
> +  TEXT_MODE_TEXT = 1,
> +  TEXT_MODE_PASSWORD = 2,
>
> ⬇️ Suggested change
>
> -  TEXT_MODE_AUTO = 0,
> -  TEXT_MODE_TEXT = 1,
> -  TEXT_MODE_PASSWORD = 2,
> +  TEXT_MODE_TEXT = 0,
> +  TEXT_MODE_PASSWORD = 1,
>
> ------------------------------
>
> In tests/component_tests/text/test_text.yaml
> <https://github.com/esphome/esphome/pull/5336#discussion_r1369520134>:
>
> > @@ -0,0 +1,31 @@
> +esphome:
>
> These tests are not used by CI.
> Please add to one of the test yaml files in the tests folder.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esphome/esphome/pull/5336#pullrequestreview-1693805000>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFZCHZU7XZAWYO4TVWGY3TYA4XFNAVCNFSM6AAAAAA4JEEEIOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMOJTHAYDKMBQGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jesserockz commented:
> I'm impressed by the thoroughness of the review and attention to code
quality!

When given the time to sit there and without distractions, its really nice to be able to sit there and review code from top to bottom.

Hopefully we can get this into the next ESPHome release, I might be able to get it into the HA November release if I work on that side tomorrow as the beta starts tomorrow for that.

nielsnl68 commented:
I will give you a peace of 🍰 when this will be finally in 👯‍♂️

jesserockz committed:
Merge branch 'dev' into pr/EternityForest/5336

jesserockz committed:
Remove AUTO and regenerate

jesserockz committed:
Remove AUTO

jesserockz committed:
Remove unnecessary file

jesserockz committed:
Call validate

jesserockz committed:
Update tests

jesserockz committed:
Fix min and max
Fix setting pattern
Use max length for restore data length
Dont set initial_value as empty string

jesserockz committed:
Add lengths and pattern hash to preference key to invalidate data on changes
Dont publish empty string as initial

jesserockz committed:
Fix test

jesserockz committed:
Reorder fields and remove unique_id

jesserockz committed:
Hide password values

jesserockz committed:
Add back unique_id (but dont use it)

jesserockz committed:
Remove testing log line

jesserockz committed:
Add to test2
